constexpr check if type is loggable

Is it possible to check (in a constexpr context) if a given type is serializable by binlog? Motivating example
```cpp
template<typename T>
void Log(T && t)
{
   if constexpr(binlog::is_serializable<T>)
   {
      BINLOG_INFO("{}", t);
   }
   else
   {
      BINLOG_INFO("Non serializable type t: {}", boost::typeindex::type_id<T>().pretty_name());
   }
}
```

```bash
include/mserialize/detail/Tag.hpp:31:5: error: static_assert failed due to requirement 'always_false<SNIP>::value' "T has no associated tag"
```
`mserialize::detail::is_serializable<TEvent>::value` is returning true but then the above static assert is failing? Does is_serializable not check for tags, is there a second constexpr check that needs to be done to guarantee a type will pass into binlog?

--- a/include/binlog/adapt_stdvariant.hpp
+++ b/include/binlog/adapt_stdvariant.hpp
@@ -55,6 +55,13 @@ struct CustomSerializer<std::variant<T...>>
 template <typename... T>
 struct CustomTag<std::variant<T...>>
 {
+  // Make this type not-constructible, if some Tag<T> is not constructible
+  std::conditional_t<
+    detail::conjunction<detail::has_tag<T>...>::value,
+    std::true_type,
+    mserialize::detail::BuiltinTag<void>
+  > tag_guard;
+
   static constexpr auto tag_string()
   {
     return cx_strcat(

--- a/include/mserialize/make_template_tag.hpp
+++ b/include/mserialize/make_template_tag.hpp
@@ -78,12 +78,25 @@
 // MSERIALIZE_UNTUPLE((Tuple<A,B,C>)) = Tuple<A,B,C>
 // MSERIALIZE_STRINGIZE_LIST(Tuple<A,B,C>) ~= "Tuple<A,B,C>"
 //
+// tag_guard: to make has_tag report correctly if T<A> has a tag or not,
+// (that depends on the template parameter), we first check if every member
+// has a tag (see conjunction), then either create a constructible member
+// (true_type, if every member has a tag), or a not-constructible member,
+// (BuiltinTag), making this spec also non-constructible - has_tag
+// checks constructibility.
 #define MSERIALIZE_MAKE_TEMPLATE_TAG(TemplateArgs, ...)           \
   namespace mserialize {                                          \
   template <MSERIALIZE_UNTUPLE(TemplateArgs)>                     \
   struct CustomTag<MSERIALIZE_UNTUPLE(MSERIALIZE_FIRST(__VA_ARGS__)), void> \
   {                                                               \
     using T = MSERIALIZE_UNTUPLE(MSERIALIZE_FIRST(__VA_ARGS__));  \
+    std::conditional_t<                                           \
+      mserialize::detail::conjunction<                            \
+        MSERIALIZE_FOREACH(MSERIALIZE_HAS_TAG, _, __VA_ARGS__)    \
+      std::true_type>::value,                                     \
+      std::true_type,                                             \
+      mserialize::detail::BuiltinTag<void>                        \
+    > tag_guard;                                                  \
     static constexpr auto tag_string()                            \
     {                                                             \
       return cx_strcat(                                           \
@@ -107,5 +120,9 @@
   /**/
 
 #define MSERIALIZE_STRINGIZE_LIST_I(_, a) "," #a
+
+#define MSERIALIZE_HAS_TAG(_, m)                                                      \
+  mserialize::detail::has_tag<decltype(mserialize::serializable_member_type(&T::m))>, \
+  /**/
 
 #endif // MSERIALIZE_MAKE_TEMPLATE_TAG_HPP

--- a/test/integration/LoggingVariants.cpp
+++ b/test/integration/LoggingVariants.cpp
@@ -91,3 +91,11 @@ int main()
 
   binlog::consume(std::cout);
 }
+
+struct Adapted {};
+BINLOG_ADAPT_STRUCT(Adapted)
+
+struct NotAdapted {};
+
+static_assert(mserialize::detail::has_tag<std::variant<Adapted>>::value, "");
+static_assert(!mserialize::detail::has_tag<std::variant<NotAdapted>>::value, "");

--- a/test/unit/mserialize/tag.cpp
+++ b/test/unit/mserialize/tag.cpp
@@ -188,3 +188,23 @@ MSERIALIZE_MAKE_DERIVED_STRUCT_TAG(Derived2, (Derived1))
 static_assert(mserialize::tag<Derived2>() == "{Derived2`'{Derived1`'{Base2`'{Base1`a'i}`b'i}`'{Base3`c'[c}`d'i`e'i}}", "");
 
 #endif // _WIN32
+
+// test has_tag
+
+// This is a private method, but still useful sometimes,
+// e.g: user wants to check if a template with a random member is loggable or not.
+
+struct Adapted{};
+MSERIALIZE_MAKE_STRUCT_TAG(Adapted)
+
+struct NotAdapted{};
+
+template <typename Nested>
+struct Nest
+{
+  Nested n;
+};
+MSERIALIZE_MAKE_TEMPLATE_TAG((typename Nested), (Nest<Nested>), n)
+
+static_assert(mserialize::detail::has_tag<Nest<Adapted>>::value, "");
+static_assert(!mserialize::detail::has_tag<Nest<NotAdapted>>::value, "");

